src/jit/optimizer.cpp

```diff
@@ -2834,365 +2829,373 @@ void Compiler::optUnrollLoops()
+        {
+            continue;
+        }
 
-            /* Locate the pre-header and initialization and increment/test statements */
+        noway_assert(init->gtOper == GT_STMT);
+        init = init->gtStmt.gtStmtExpr;
+        noway_assert(test->gtOper == GT_STMT);
+        test = test->gtStmt.gtStmtExpr;
+        noway_assert(incr->gtOper == GT_STMT);
+        incr = incr->gtStmt.gtStmtExpr;
 
-            phdr = head->bbTreeList;
-            noway_assert(phdr);
-            loop = bottom->bbTreeList;
-            noway_assert(loop);
+        // Don't unroll loops we don't understand.
+        if (incr->gtOper != GT_ASG)
+        {
+            continue;
+        }
+        incr = incr->gtOp.gtOp2;
 
-            init = head->lastStmt();
-            noway_assert(init && (init->gtNext == nullptr));
-            test = bottom->lastStmt();
-            noway_assert(test && (test->gtNext == nullptr));
-            incr = test->gtPrev;
-            noway_assert(incr);
+        /* Make sure everything looks ok */
+        if ((init->gtOper != GT_ASG) || (init->gtOp.gtOp1->gtOper != GT_LCL_VAR) ||
+            (init->gtOp.gtOp1->gtLclVarCommon.gtLclNum != lvar) || (init->gtOp.gtOp2->gtOper != GT_CNS_INT) ||
+            (init->gtOp.gtOp2->gtIntCon.gtIconVal != lbeg) ||
 
-            if (init->gtFlags & GTF_STMT_CMPADD)
-            {
-                /* Must be a duplicated loop condition */
-                noway_assert(init->gtStmt.gtStmtExpr->gtOper == GT_JTRUE);
+            !((incr->gtOper == GT_ADD) || (incr->gtOper == GT_SUB)) || (incr->gtOp.gtOp1->gtOper != GT_LCL_VAR) ||
+            (incr->gtOp.gtOp1->gtLclVarCommon.gtLclNum != lvar) || (incr->gtOp.gtOp2->gtOper != GT_CNS_INT) ||
+            (incr->gtOp.gtOp2->gtIntCon.gtIconVal != iterInc) ||
 
-                dupCond = true;
-                init    = init->gtPrev;
-                noway_assert(init);
-            }
-            else
-            {
-                dupCond = false;
-            }
+            (test->gtOper != GT_JTRUE))
+        {
+            noway_assert(!"Bad precondition in Compiler::optUnrollLoops()");
+            continue;
+        }
 
-            /* Find the number of iterations - the function returns false if not a constant number */
+        /* heuristic - Estimated cost in code size of the unrolled loop */
 
-            if (!optComputeLoopRep(lbeg, llim, iterInc, iterOper, iterOperType, testOper, unsTest, dupCond, &totalIter))
-            {
-                continue;
-            }
+        loopCostSz = 0;
 
-            /* Forget it if there are too many repetitions or not a constant loop */
+        block         = head->bbNext;
+        auto tryIndex = block->bbTryIndex;
 
-            if (totalIter > iterLimit)
+        for (;; block = block->bbNext)
+        {
+            if (block->bbTryIndex != tryIndex)
             {
-                continue;
+                // Unrolling would require cloning EH regions
+                goto DONE_LOOP;
```

Comment:
I got rid of the fixpoint iteration so we won't revisit the same loop twice anymore.